prediction_market_agent/agents/microchain_agent/blockchain/contract_class_converter.py

```diff
@@ -0,0 +1,184 @@
+    AbiItemStateMutabilityEnum,
+    AbiItemTypeEnum,
+    ABIMetadata,
+)
+from prediction_market_agent.utils import APIKeys
+
+TYPE_MAPPING: dict[str, Tuple[str, Any]] = {
+    "address": (
+        "str",
+        Web3.to_checksum_address("0xe91d153e0b41518a2ce8dd3d7944fa863463a97d"),
+    ),  # Solidity address to Python string
+    "uint": ("int", 1),  # Solidity unsigned integer to Python int
+    "uint8": ("int", 1),  # Solidity unsigned integer to Python int
+    "uint256": ("int", 1),  # Solidity unsigned integer (256-bit) to Python int
+    "int": ("int", 1),  # Solidity integer to Python int
+    "bool": ("bool", True),  # Solidity boolean to Python bool
+    "string": ("str", "test string"),
+    # Add more mappings as needed
+}
+
+
+class FunctionWithKeys(Function):
+    def __init__(self, keys: APIKeys) -> None:
+        self.keys = keys
+        super().__init__()
+
+
+class ClassFactory:
+    @staticmethod
+    def create_class(
+        class_name: str,
+        base_classes: Tuple[type],
+        attributes: dict[str, Any] | None = None,
+    ) -> type:
+        if attributes is None:
+            attributes = {}
+
+        new_class = type(class_name, base_classes, attributes)
+        return new_class
+
+
+class ContractClassConverter:
+    """Class responsible for reading a smart contract on Gnosis Chain and converting its functionalities into Python classes."""
+
+    def __init__(self, contract_address: ChecksumAddress):
+        # For caching requests of the same contract
+        self.contract_address = contract_address
+
+    def fetch_from_blockscout(self) -> dict[str, Any]:
+        r = requests.get(
+            f"https://gnosis.blockscout.com/api/v2/smart-contracts/{self.contract_address}"
+        )
+        r.raise_for_status()
+        data: dict[str, Any] = r.json()
+        return data
+
+    def get_abi(self) -> list[ABIMetadata]:
+        data = self.fetch_from_blockscout()
+        contract_abi = data["abi"]
+        # We extract only functions, not events
+        return [
+            ABIMetadata.parse_obj(abi_item)
+            for abi_item in contract_abi
+            if abi_item["type"] == "function"
+        ]
+
+    def get_source_code(self) -> str:
+        data = self.fetch_from_blockscout()
+        source_code = data["source_code"]
+        return str(source_code)
+
+    def generate_microchain_class_from_abi_item(
+        self,
+        abi_item: ABIMetadata,
+        contract: ContractOnGnosisChain,
+        summaries: Summaries,
+    ) -> Tuple[AbiItemStateMutabilityEnum | None, type | None]:
+        if abi_item.type != AbiItemTypeEnum.function:
+            return None, None
+
+        # If type mapping fails, we log and fail gracefully. Note that structs as input- or output args are not supported.
+        for input in abi_item.inputs:
+            if not TYPE_MAPPING.get(input.type, None):
+                logger.info(f"Type mapping has failed. Check inputs {abi_item.inputs}")
```

Comment:
What should 'failing gracefully' look like in this context? Looks like it will hit an exception below at `TYPE_MAPPING[input.type][0]` if the type string is not in the mapping dict

Reply:
Good point, it was supposed to return None if type mapping fails.
I expect this type mapping to fail e.g. for structs, and that's fine in the beginning.

Reply:
Oh sorry, you return `None, None` now. lgtm